### ISSUE TITLE
Clarify sub-typing behavior in Q#

### ISSUE THREAD

**Issue by vadym-kl** [microsoft/qsharp-compiler#433]
**Is your feature request related to a problem? Please describe.**
Consider the following snippet of code:

```qsharp
function fooA( a : (Unit => Unit is Adj) ) : Unit {}
function foo( a : (Unit => Unit) ) : Unit {}
function Bar() : Unit {
    let x = [foo, fooA];
    let z = [(1, foo), (2, fooA)];
    let y = [(foo, foo), (fooA, fooA)];
    // (QS5002) Array items must have a common base type.
    let w = [[foo], [fooA]]; // QS5002 error
}
```
Expressions `let x ...`, `let y ...`, `let z ...` compile without an error, and `let w=...` causes an error. 
It seems like `(Unit => Unit)` is a supertype of `(Unit => Unit is Adj)`, however 
`(Unit => Unit)[]` and  `(Unit => Unit is Adj)[]` have no common base type. 

**Describe the solution you'd like**
I would like the code above to compile without an error. 

**Describe alternatives you've considered**
Update the documentation at https://docs.microsoft.com/en-us/quantum/user-guide/language/types#operation-and-function-types to clarify that above is intended behavior

**Comment by bettinaheim** [comment]
@vadym-kl Indeed, Q# arrays at this time are invariant, rather than covariant as you may expect them to be. This is by design and may be relaxed in the future. Please let me know if you have any further questions.

**Comment by bettinaheim** [comment]
I'll transfer this to the docs repo, since it indeed looks like we don't (no longer?) explicitly state that arrays are invariant.

**Comment by bettinaheim** [comment]
I can't transfer, so I recreated the issue on our docs repo.

**Comment by vadym-kl** [comment]
@bettinaheim, should this request be left open if array invariance may be relaxed in the future?

**Comment by bettinaheim** [comment]
@vadym-kl I can rephrase the title and keep it open until we can make an informed call on this. It will be a while before we could possibly support that though, if at all.

**Comment by alan-geller** [comment]
I'm going to transfer this issue to the qsharp-language repo, since that's now the proper place for language suggestions.
